### ISSUE TITLE
Bug using stat_pvalue_manual : Brackets don't correspond to data

### ISSUE THREAD

**Issue by Minh-AnhHuynh** [kassambara/ggpubr#477]
Hello,

I am trying to add my own pvalues from a Tukey test since stat_compare_means does not allow us to do post-hoc anova testing. I am referring myself to the [git issue there](https://github.com/kassambara/ggpubr/issues/102) to use stat_pvalue_manual to add to a ggpubr plot.

However brackets from stat_pvalue_manual, taken from the result of rstatix::tukey_hsd() do not actually correspond to the actual results in the tibble format.

Here's a reproductible example.
```
library(ggpubr)
library(rstatix)
data <- structure(list(sample = c(
  "PARDAX", "0A4HX8", "PAUYTT", "PAKXLD",
  "PAMTCM", "PAPWWC", "0A4HXS"
), immune = c(
  -25.0149700950069,
  31.8765982455667, -291.974237078435, -834.658161729368, -739.769040781408,
  469.410793948452, 577.65693757783
), Group = c(
  "Medium", "Medium",
  "Low", "Low", "Low", "Low", "High"
)), row.names = c(NA, -7L), class = c(
  "tbl_df",
  "tbl", "data.frame"
))
stat_tukey <- tukey_hsd(data, immune ~ Group) %>% add_xy_position()
print(stat_tukey)
#> # A tibble: 3 × 13
#>   term  group1 group2 null.value estimate conf.low conf.high p.adj p.adj.signif
#>   <chr> <chr>  <chr>       <dbl>    <dbl>    <dbl>     <dbl> <dbl> <chr>       
#> 1 Group High   Low             0    -927.   -2981.     1127. 0.342 ns          
#> 2 Group High   Medium          0    -574.   -2825.     1676. 0.664 ns          
#> 3 Group Low    Medium          0     353.   -1239.     1944. 0.728 ns          
#> # … with 4 more variables: y.position <dbl>, groups <named list>, xmin <dbl>,
#> #   xmax <dbl>
ve <-
  ggviolin(
    data,
    "Group",
    "immune",
    ylab = "Immune Score",
    fill = "Group",
    order = c("Low", "Medium", "High")
  ) +
  stat_compare_means(
    method = "anova",
    label.x.npc = "left",
    label.y.npc = "top"
  ) +
  stat_pvalue_manual(stat_tukey,
    label = "p.adj",
    step.increase = 0.1
  )

```

![image](https://user-images.githubusercontent.com/96783020/176642186-619520d7-29ce-4979-92af-85e0008e97ac.png)

As you can see, the bracket between "Low" and "Medium" group correspond to p = 0.342 BUT the corresponding group in the tibble clearly indicates that the value is for "High" vs "Low" group!
Apologies for the hideous image.

My guess is that stat_pvalue_manual processes the brackets in a linear fashion (from first row to last_row) and doesn't read the group, which is strange.

I may be using this wrong, but if I'm not mistaken it is what is proposed in  [this guide on datanovia.](https://www.datanovia.com/en/blog/how-to-add-p-values-onto-basic-ggplots/#create-plots-with-significance-levels)

Thanks,
Minh-Anh

**Comment by kassambara** [comment]
To avoid this issue, the easiest solution is to modify your data by defining the grouping variable level order, using R code like this:

`my_data[["my_grouping_variable"]] = factor( my_data[["my_grouping_variable"]], levels = c("grp1", "grp2", "grp3"))`

Then, use the data to do statistical tests and plots.

See example below:

``` r
suppressPackageStartupMessages(library(ggpubr))
suppressPackageStartupMessages(library(rstatix))

data <- structure(list(sample = c(
  "PARDAX", "0A4HX8", "PAUYTT", "PAKXLD",
  "PAMTCM", "PAPWWC", "0A4HXS"
), immune = c(
  -25.0149700950069,
  31.8765982455667, -291.974237078435, -834.658161729368, -739.769040781408,
  469.410793948452, 577.65693757783
), Group = c(
  "Medium", "Medium",
  "Low", "Low", "Low", "Low", "High"
)), row.names = c(NA, -7L), class = c(
  "tbl_df",
  "tbl", "data.frame"
))

#  Specify the order of grouing variable levels
data[["Group"]] <- factor(data[["Group"]], c("Low", "Medium", "High"))


# Statiistical tests
stat_tukey <- tukey_hsd(data, immune ~ Group) %>% add_xy_position()
stat_tukey
#> # A tibble: 3 × 13
#>   term  group1 group2 null.value estimate conf.low conf.…¹ p.adj p.adj…² y.pos…³
#>   <chr> <chr>  <chr>       <dbl>    <dbl>    <dbl>   <dbl> <dbl> <chr>     <dbl>
#> 1 Group Low    Medium          0     353.   -1239.   1944. 0.728 ns         643.
#> 2 Group Low    High            0     927.   -1127.   2981. 0.342 ns         741.
#> 3 Group Medium High            0     574.   -1676.   2825. 0.664 ns         840.
#> # … with 3 more variables: groups <named list>, xmin <dbl>, xmax <dbl>, and
#> #   abbreviated variable names ¹​conf.high, ²​p.adj.signif, ³​y.position


# Plot
ggviolin(
  data,
  "Group",
  "immune",
  ylab = "Immune Score",
  fill = "Group",
  trim = TRUE
) +
stat_compare_means(
  method = "anova",
  label.x.npc = "left",
  label.y = 1100
) +
stat_pvalue_manual(stat_tukey,
                   label = "p.adj",
                   step.increase = 0.1
)
#> Warning: Groups with fewer than two data points have been dropped.
```

![](https://i.imgur.com/JpHRIeO.png)

**Comment by Minh-AnhHuynh** [comment]
Thank you for the answer. I'll keep that in mind when needed.
Another thing I found out is that `add_xy_position()` is not always desirable.

`add_xy_position()` adds xmin and xmax in order to place the brackets, but those do NOT correspond to the appropriate groups in violin plot. Instead, I used `add_y_position()` and `xmin` and `xmax``url` is referred to `group1 `and `group2 `by default by `stat_pvalue_manual()`. I hope it helps someone!
